### DISCUSSION TITLE
Packaging Streamlit web app into an Electron desktop app: non-pure Python packages

### DISCUSSION THREAD

**Discussion by lechtec** [whitphx/stlite#388] [Q&A]
I'm a big fan of the idea of an automated service to bundle up a Streamlit app developed in Python in an Electron desktop app. This  provides a great way for data scientists and developers to easily share their dashboards to non-technical crowds, without the need of having to figure out hosting for the dashboard on a web server. 

### My question
My question relates to `stlite-cli`. In general, is there any way to export non-pure Python packages into an Electron desktop app? If so, what does this process require? If this method already exists, does anyone know of implementations that have achieved this before? I am a novice in Electron, and I haven't found any straightforward resources for doing that. 

Thank you!

**Comment by whitphx** [comment]
Thank you, I totally agree on the possibility of such an automated bundling service! That's on the roadmap, but takes time too. So please be patient until it :)

For your question,
do you mean that you want to use non-pure Python packages such as NumPy within the stlite Electron app?
If so, you can simply use the `-r` option of the dump command that you can find in the [document](https://github.com/whitphx/stlite/tree/main/packages/desktop-cli#readme) by grepping `-r`, for example, `npm run dump streamlit_app -- -r numpy`
[This sample project](https://github.com/whitphx/stlite-desktop-example) uses this method to install matplotlib (See https://github.com/whitphx/stlite-desktop-example/blob/b7c531c46aab9e612039cc15dfcac83f33629c33/.github/workflows/build.yml#L30)

**[Update on 0.16.0]**
The command signature has been changed so that the `-r` option is now for specifying the requirements file like `pip`. For the details, see the README.

However, only the C packages compiled for the Pyodide (Wasm) runtime are installable like NumPy, and there are some unavailable C extensions on stlite such as TensorFlow. (https://github.com/whitphx/stlite#limitations)

>  I am a novice in Electron, 

FYI, TBH, I'm too.
In the stlite desktop app projects, [electron-builder](https://www.electron.build/) is directly used (see the `dist` and `pack` commands in the `package.json` of the project template https://github.com/whitphx/stlite/tree/main/packages/desktop-cli#readme) and I actually don't care what happens behind 😅
